resources/benchmark-runner.mjs

```diff
@@ -160,112 +176,107 @@ export class BenchmarkRunner {
-            this._appendFrame();
-            this._page = new Page(this._frame);
-            return state.prepareCurrentSuite(this._page).then(
-                () => this._runTestAndRecordResults(state))
+    async _runSuite(suite)
+    {
+        await this._prepareSuite(suite);
+        for (const test of suite.tests) {
+            await this._runTestAndRecordResults(suite, test)
         }
+    }
 
-        return this._runTestAndRecordResults(state);
+    async _prepareSuite(suite)
+    {
+        return new Promise((resolve) => {
+            const frame = this._page._frame;
+            frame.onload = async () => {
+                await suite.prepare(this._page)
+                resolve();
+            }
+            frame.src = 'resources/' + suite.url;
+        });
     }
 
-    runAllSteps(startingState) {
-        const nextCallee = this.runAllSteps.bind(this);
-        this.step(startingState).then(nextState => {
-            if (nextState)
-                nextCallee(nextState);
+    async _runTestAndRecordResults(suite, test)
+    {
+        return new Promise((resolve) => {
+            if (this._client && this._client.willRunTest)
+                this._client.willRunTest(suite, test);
+
+             window.requestAnimationFrame(
+                this._runTest.bind(this, suite, test, this._page, resolve));
         });
     }
 
-    runMultipleIterations(iterationCount) {
-        const self = this;
-        let currentIteration = 0;
-
-        this._runNextIteration = () => {
-            currentIteration++;
-            if (currentIteration < iterationCount)
-                self.runAllSteps();
-            else if (this._client && this._client.didFinishLastIteration)
-                this._client.didFinishLastIteration();
-        }
+    // This function ought be as simple as possible. Don't even use Promise.
+    _runTest(suite, test, page, test_done_callback)
+    {
+        this._writeMark(suite.name + '.' + test.name + '-start');
+        const startTime = window.performance.now();
+        test.run(page);
+        const endTime = window.performance.now();
+        this._writeMark(suite.name + '.' + test.name + '-sync-end');
 
-        if (this._client && this._client.willStartFirstIteration)
-            this._client.willStartFirstIteration(iterationCount);
+        const syncTime = endTime - startTime;
+        this._measureTestAsyncTime(suite, test, syncTime, test_done_callback)
+    }
 
-        self.runAllSteps();
+    _measureTestAsyncTime(suite, test, syncTime, test_done_callback)
+    {
+        const startTime = window.performance.now();
+        setTimeout(() => {
+            const endTime = window.performance.now();
+            this._writeMark(suite.name + '.' + test.name + '-async-end');
+
+            const asyncTime = endTime - startTime;
+            window.requestAnimationFrame(
+                this._recordTestResults.bind(this, suite, test, syncTime, asyncTime, test_done_callback));
```

Comment:
I got into the habit of using bind instead of arrow-function to prevent potential context-variable leaks. (can be changed back of course)